zvmsdk/volumeop.py

```
@@ -1755,26 +1754,38 @@ def _add_disks(self, fcp_list, assigner_id, target_wwpns, target_lun,
                                       target_lun, multipath, os_version,
                                       mount_point)
 
-    def _rollback_dedicated_fcp(self, fcp_list, assigner_id,
-                                all_fcp_list=None):
+    def _rollback_dedicated_fcp(self, fcp_list, assigner_id):
+        """
+        Rollback for the following completed operations:
+        1. operations on z/VM done by _dedicate_fcp()
+           i.e. undedicate FCP device from assigner_id
+        2. operations on FCP DB done by get_volume_connector()
+           i.e. unreserve FCP device and cleanup tmpl_id from FCP DB
+
+        :param fcp_list: (list) a list of FCP devices
+        :param assigner_id: (str) the userid of the virtual machine
+        :return: None
+        """
         with database.get_fcp_conn():
-            # fcp param should be a list
+            # save the FCP device whose connections == 0
+            no_connection_fcps = list()
+            # Operation on z/VM:
+            # undedicate FCP device from assigner_id
             for fcp in fcp_list:
                 with zvmutils.ignore_errors():
-                    LOG.info("Rolling back dedicated FCP: %s" % fcp)
-                    connections = self.fcp_mgr.decrease_fcp_usage(
-                        fcp, assigner_id)
+                    connections = self.db.get_connections_from_fcp(fcp)
                     if connections == 0:
+                        no_connection_fcps.append(fcp)
                         self._undedicate_fcp(fcp, assigner_id)
```

for this undedicate, I think we should ignore errors like "FCP device not exist", right?

---

The `with zvmutils.ignore_errors()` ignore any Exception, 
so as to ensure _undedicate_fcp() to be run on the last FCP device in fcp_list,
even if _undedicate_fcp() raise any exception on the previous FCP devices in fcp_list